### ISSUE TITLE
The Mega Hash of Global Boards

### ISSUE THREAD

**Issue by manumei** [manudelp/ultimate-tic-tac-toe#30]
Importante! Si queres ver la mejor move por board directo, maybe hay que retrieve la last move made, asi vez cual es tu board_to_play 
Porque si queres ver la mejor, tenes que ver que con el btp actual puedas jugarla
Los 10 board_to_play posibles son: [0, 0], [0, 1], [0, 2], [1, 0], [1, 1], [1, 2], [2, 0], [2, 1], [2, 2], None

## Ideas for Faster Hashing: 
Aclaracion, si usas esto como get_board_balance
Hace hashed_balance = get_board_balance(board)
If balance is not None, return hashed_balance
else, return self.boardBalance(board)

☆ Hash Combinations of Valid Local Boards. **See more in generate_legal_boards function in generate_hashes.py!** 
This works since valid local boards comb number is much lower than 3**9. Illegal boards include boards where both players are winners, and boards where a player has won from more than two 3-in-lines at the same time. There is only one possible combination of winning with three 3-in-lines, the one in this picture (and its other 3 rotations of course), add it by hand, and then limit it only to winning lines < 3. 
![Image](https://github.com/user-attachments/assets/3b74d148-a650-4e25-893c-095dc1bad7a1)

☆ **Huge one!!** All won local boards are the same to me. I don't need to check for the 5000 different types of local won boards. So I can generate my Local Boards from a set that contains only the non-won local boards, and then one local board won by player=1 & one local board won by player=-1, and that's it! Since all local boards that are already won, are the same to me.
In fact, same goes to the fully drawn boards, just narrow them down to a single board. This is gigantic
_Disclaimer: Para hacer esto tendrías que, al principio de cada turno, editar el tablero posta que vayas a usar, reemplazandole todos los won local boards y drawn local boards, por las variables normalizadas que uses. Sino despues cuando busques en el hash, no vas a encontrar matches porque solo hasheaste 1 kind de tablero ganado, pero en tu board esta ganado de otra forma. Normaliza todos los tableros ganados y empatados para que se vean siempre igual. Y al simular las jugadas en el minimax tambien, haces despues de cada simulacion a un local board, if winner(local_board) = 1: board[r, c] = my_normalized_wonby1_board y respectivamente para won by -1 when winner=-1._

☆ You only need to add 1s and -1s, and the default board can be np.zeros(), maybe this can help go to base2 instead of base3

☆ Board Symmetry is important! Each board can be rotated in 4 directions and reflected in 4 as well (total of 8 symmetrical pairs per board included), reducing the amount greatly (may be valid only for the global board as rotating and reflecting local boards differently within the same global is technically different... but if you're only using heuristic function on it, then it's likely your heuristic won't see the difference, granted, this comes from a flaw in the heuristic that's very hard to overcome)

☆ Global Board amount of 1s and -1s has to be the same, or difference of 1 between each other.
Such that |count(1) - count(-1)| ≤ 1. This is also a limit for the possibilities of boards to evaluate

☆ The Center Local Board (1, 1) cannot have an empty center, meaning (1, 1, 1, 1) cannot be 0

☆ There have to be less than 79 empty available tiles, (before that, most of it will follow along similar lines)

☆ There have to be more than 10 empty available tiles, otherwise I can already handle without board balance

☆ Local Boards can't have a piece difference between 1s and -1s larger than 4

## Once the Hash is Ready
When Board Balance is good, start looking into pre-hashing overall global_board heuristic values
And maybe even pre-hashing a lot of global_board moves directly, after calculating them with minimax, maybe...
Basically, when I say "a lot" is cause I mean at least millions actually

## Old Idea, for hashing the 9x9 directly
    - El tablero es un np.zeros, solo tenes que agregarle 2^81 at most (1s & -1s)
    - El tablero si o si tiene o la misma cantidad de 1s y -1s, o 1 mas de 1s que de -1s 
    - (no se si esto lo facilita o lo complica), pero el tablero solo es valido con las reglas de board_to_play
    - symmetry! divide by 4  symmetry
